src/uu/tail/src/args.rs

```diff
@@ -362,6 +362,72 @@ fn parse_num(src: &str) -> Result<Signum, ParseSizeError> {
+/// However, internally the parsed [`Duration`] is capped at `seconds == u64::MAX`, `nanos (max) ==
+/// .999999999` and `nanos (min if not 0) == .000000001`. Infinity values like `inf`, `+infinity`
+/// etc. are valid input and resolve to `Duration::MAX`.
+///
+/// # Arguments
+///
+/// * `src` - A string slice that contains a f64 value
+///
+/// # Errors
+///
+/// This function will return an error when parsing fails (See also [`f64::from_str`]) or the `src`
+/// was negative (-0.0 counts as not negative).
+///
+/// # Examples
+///
+/// ```ignore
+/// use std::time::Duration;
+///
+/// let duration = parse_duration("+1.09e1").unwrap();
+/// assert_eq!(duration, Duration::new(10, 900_000_000));
+/// ```
+///
+/// We accept the same max values like gnu's tail on the command line by parsing into `f64`.
+/// Internally we can only use [`Duration::MAX`] and cap the seconds at [`u64::MAX`]. By doing the
+/// conversion like below, we avoid the panic when using [`Duration::from_secs_f64`] because the
+/// value is too big. The [`Duration::try_from_secs_f64`] isn't yet stable and would also lack the
+/// ability to use the max value in case of an error.
+///
+/// [`f64::from_str`]: https://doc.rust-lang.org/std/primitive.f64.html#method.from_str
+fn parse_duration(src: &str) -> Result<Duration, String> {
+    let trimmed = src.trim();
+    match src.parse::<f64>() {
+        // We're working here with the string representation instead of num.fract() to avoid further
+        // floating point precision problems where we can.
+        Ok(num) if num.is_finite() && num >= 0.0 => match num.to_string().split_once('.') {
+            Some((_, fract)) => {
+                let mut multi = 100_000_000;
+                let mut nanos: u32 = 0;
+                for c in fract.chars().take(9) {
+                    nanos += c.to_digit(10).unwrap() * multi;
+                    multi /= 10;
+                }
+                Ok(Duration::new(num as u64, nanos))
+            }
+            None if num <= u64::MAX as f64 => Ok(Duration::new(num as u64, 0)),
+            None => Ok(Duration::new(u64::MAX, 999_999_999)),
+        },
+        Ok(num)
+            if num.is_infinite()
+                && (trimmed.eq_ignore_ascii_case("inf")
+                    || trimmed.eq_ignore_ascii_case("+inf")
+                    || trimmed.eq_ignore_ascii_case("infinity")
+                    || trimmed.eq_ignore_ascii_case("+infinity")) =>
+        {
+            Ok(Duration::MAX)
+        }
+        Ok(num) if num.is_sign_negative() => Err(String::from("Number was negative")),
+        // positive infinite (if src > f64::MAX) or NaN. In case of positive infinite we need to
+        // match gnu's tail behavior and return an error although it may be a nice feature to
+        // interpret this result as a valid (maximum) Duration.
+        Ok(_) => Err(String::from("Not a number")),
```

Comment:
Setting Duration::MAX makes sense to me here if it is a valid positive number, but that's a bit tricky to check (especially with an exponent). So cool idea!

Reply:
> Setting Duration::MAX makes sense to me here if it is a valid positive number

Ok, I think checking for `sign_positive` in the match arm

```rust
Ok(num) if num.is_infinite() ...
```

should do the trick.

Reply:
I haven't changed this because of your comment above.
> Oh I see, I assumed it would be Err if it was more than f64::MAX, but that makes sense. If you could add a quick comment explaining that in the code that would be great!